core-bundle/src/Resources/contao/library/Contao/Search.php

```diff
@@ -572,4 +565,38 @@ public static function getInstance()
 
 		return static::$objInstance;
 	}
+
+	/**
+	 * @param string $strUrlA
+	 * @param string $strUrlB
+	 *
+	 * @return int negative if $strUrlA is more canonical, positive if $strUrlB is more canonical
+	 */
+	private static function compareUrls($strUrlA, $strUrlB)
+	{
+		if (strpos($strUrlA, '?') === false && strpos($strUrlB, '?') !== false)
+		{
+			return -1;
+		}
+
+		if (strpos($strUrlA, '?') !== false && strpos($strUrlB, '?') === false)
+		{
+			return 1;
+		}
+
+		$slashCountA = substr_count(explode('?', $strUrlA)[0], '/');
+		$slashCountB = substr_count(explode('?', $strUrlB)[0], '/');
+
+		if ($slashCountA !== $slashCountB)
+		{
+			return $slashCountA - $slashCountB;
+		}
+
+		if (\strlen($strUrlA) !== \strlen($strUrlB))
+		{
+			return \strlen($strUrlA) - \strlen($strUrlB);
+		}
+
+		return strcmp($strUrlA, $strUrlB);
```

Comment:
This might be more concise (+ would also limit the output to `{-1, 0, 1}` like with the native compare functions).

```suggestion
		return $slashCountA <=> $slashCountB ?: \strlen($strUrlA) <=> \strlen($strUrlB) ?: strcmp($strUrlA, $strUrlB);
```